### PR TITLE
[ConstraintSystem] Support solving expression patterns via injecting call to `~=` operator

### PR DESCRIPTION
**Pull request by xedin**
Augment the constraint solver to fallback to implicit `~=` application
when member couldn't be found for `EnumElement` patterns because
`case` statement should be able to match enum member directly, as well
as through an implicit `~=` operator application.

### CODE REVIEW

File: test/Constraints/result_builder_diags.swift [hunk 1]
```diff
@@ -443,8 +443,8 @@ func getSomeEnumOverloaded(_: Int) -> E2 { return .b(0, nil) }
 func testOverloadedSwitch() {
   tuplify(true) { c in
     // FIXME: Bad source location.
```
Comment (LucianoPAlmeida) [inline test/Constraints/result_builder_diags.swift:445]: Does this comment still make sense now that source loc seems to be correct?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -8629,6 +8629,69 @@ fixMemberRef(ConstraintSystem &cs, Type baseTy,
   return nullptr;
 }
 
+/// Convert the given enum element pattern into an expression pattern
+/// and synthesize ~= operator application to find the type of the
+/// element.
+static bool inferEnumMemberThroughTildeEqualsOpeator(
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:8635]: Typo? :)
```suggestion
static bool inferEnumMemberThroughTildeEqualsOperator(
```

File: test/Constraints/result_builder_diags.swift (hunk 1 above)
Comment (xedin) [reply test/Constraints/result_builder_diags.swift:445]: Ah yeah, that doesn't make sense anymore!

Comment (xedin) [reply test/Constraints/result_builder_diags.swift]: Removed.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Fixed, thanks!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
Going to re-run macOS before merging just in case anything changed...

**Pull request merged**